File: csharp-dotnet/tests/Architecture.Shell.Cqrs.Tests/BehaviorMatcherTests.cs
using Architecture.Shell.Cqrs;

namespace Architecture.Shell.Cqrs.Tests;

/// <summary>
/// UT-007: BehaviorMatcher Type Guards
/// </summary>
public class BehaviorMatcherTests
{
    // UT-007: T061b
    [Fact]
    public void Should_MatchCommands_When_IsCommandGuardUsed()
    {
        // Given: BehaviorMatcher configured to match only commands
        var commandMatcher = new CommandOnlyMatcher();
        var queryMatcher = new QueryOnlyMatcher();

        var command = new TestCommand();
        var query = new TestQuery { Id = 1 };

        // When: Matcher is evaluated against command and query
        var commandMatchesCommandMatcher = commandMatcher.Matches(command);
        var queryMatchesCommandMatcher = commandMatcher.Matches(query);
        var commandMatchesQueryMatcher = queryMatcher.Matches(command);
        var queryMatchesQueryMatcher = queryMatcher.Matches(query);

        // Then: Command matcher should only match commands
        commandMatchesCommandMatcher.Should().BeTrue();
        queryMatchesCommandMatcher.Should().BeFalse();

        // And: Query matcher should only match queries
        commandMatchesQueryMatcher.Should().BeFalse();
        queryMatchesQueryMatcher.Should().BeTrue();
    }

    [Fact]
    public void Should_MatchAllRequests_When_AllRequestsMatcherUsed()
    {
        // Given: BehaviorMatcher configured to match all requests
        var allMatcher = new AllRequestsMatcher();

        var command = new TestCommand();
        var query = new TestQuery { Id = 1 };

        // When: Matcher is evaluated
        var commandMatches = allMatcher.Matches(command);
        var queryMatches = allMatcher.Matches(query);

        // Then: Both should match
        commandMatches.Should().BeTrue();
        queryMatches.Should().BeTrue();
    }
}

// Matcher implementations
public class CommandOnlyMatcher : IBehaviorMatcher
{
    public bool Matches<TRequest>(TRequest request) where TRequest : IBaseRequest
    {
        return request is ICommand;
    }
}

public class QueryOnlyMatcher : IBehaviorMatcher
{
    public bool Matches<TRequest>(TRequest request) where TRequest : IBaseRequest
    {
        var requestType = typeof(TRequest);
        return requestType.GetInterfaces().Any(i =>
            i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IQuery<>));
    }
}

public class AllRequestsMatcher : IBehaviorMatcher
{
    public bool Matches<TRequest>(TRequest request) where TRequest : IBaseRequest
    {
        return true;
    }
}

File: java-spring/architecture-shell-cqrs/src/test/java/com/architecture/shell/cqrs/BehaviorMatcherTests.java
package com.architecture.shell.cqrs;

import com.architecture.core.functional.Result;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Unit tests for BehaviorMatcher Type Guards.
 * Tests follow TDD approach - written before implementation exists.
 * Naming: Should_ExpectedBehavior_When_StateUnderTest
 */
class BehaviorMatcherTests {

    /**
     * T061c: UT-007 - BehaviorMatcher Type Guards (IsCommand)
     * Given: BehaviorMatcher with IsCommand guard
     * When: Request types (commands and queries) are tested
     * Then: Should match only command types
     */
    @Test
    void should_MatchCommands_When_IsCommandGuardUsed() {
        // Given: Command-only matcher
        BehaviorMatcher commandMatcher = request -> request instanceof Command;

        // When/Then: Should match commands
        assertThat(commandMatcher.matches(new TestCommand())).isTrue();
        assertThat(commandMatcher.matches(new TestCommandWithResult())).isTrue();

        // When/Then: Should NOT match queries
        assertThat(commandMatcher.matches(new TestQuery())).isFalse();
    }

    /**
     * UT-007 - BehaviorMatcher Type Guards (IsQuery)
     * Given: BehaviorMatcher with IsQuery guard
     * When: Request types (commands and queries) are tested
     * Then: Should match only query types
     */
    @Test
    void should_MatchQueries_When_IsQueryGuardUsed() {
        // Given: Query-only matcher
        BehaviorMatcher queryMatcher = request -> request instanceof Query;

        // When/Then: Should match queries
        assertThat(queryMatcher.matches(new TestQuery())).isTrue();

        // When/Then: Should NOT match commands
        assertThat(queryMatcher.matches(new TestCommand())).isFalse();
        assertThat(queryMatcher.matches(new TestCommandWithResult())).isFalse();
    }

    /**
     * UT-007 - BehaviorMatcher Type Guards (AllRequests)
     * Given: BehaviorMatcher that matches all requests
     * When: Any request type is tested
     * Then: Should match all request types
     */
    @Test
    void should_MatchAllRequests_When_AllRequestsGuardUsed() {
        // Given: All-requests matcher
        BehaviorMatcher allRequestsMatcher = request -> request instanceof BaseRequest;

        // When/Then: Should match all request types
        assertThat(allRequestsMatcher.matches(new TestCommand())).isTrue();
        assertThat(allRequestsMatcher.matches(new TestCommandWithResult())).isTrue();
        assertThat(allRequestsMatcher.matches(new TestQuery())).isTrue();
    }

    // Test fixtures

    private static class TestCommand implements Command<Result<Void>> {
    }

    private static class TestCommandWithResult implements Command<Result<String>> {
    }

    private static class TestQuery implements Query<String> {
    }
}
